src/scanpy/metrics/_common.py

```diff
@@ -1,23 +1,85 @@
 from __future__ import annotations
 
 import warnings
+from abc import ABC, abstractmethod
+from dataclasses import InitVar, dataclass, field
 from functools import singledispatch
-from typing import TYPE_CHECKING, overload
+from typing import TYPE_CHECKING, ClassVar, overload
 
 import numpy as np
 import pandas as pd
 from scipy import sparse
 
-from .._compat import DaskArray
+from .._compat import DaskArray, fullname
 
 if TYPE_CHECKING:
     from typing import NoReturn, TypeVar
 
+    from anndata import AnnData
     from numpy.typing import NDArray
 
     T_NonSparse = TypeVar("T_NonSparse", bound=NDArray | DaskArray)
     V = TypeVar("V", bound=NDArray | sparse.csr_matrix)
 
+    _Vals = NDArray | sparse.spmatrix | DaskArray | pd.DataFrame | pd.Series
+
+
+__all__ = ["_get_graph", "_SparseMetric"]
+
+
+@dataclass
+class _SparseMetric(ABC):
+    g: sparse.csr_matrix
+    vals: InitVar[_Vals]
+    _vals: NDArray | sparse.csr_matrix | DaskArray = field(init=False)
+
+    name: ClassVar[str]
+
+    def __post_init__(self, vals: _Vals) -> None:
+        assert isinstance(type(self).name, str)
+        assert self.g.shape[0] == self.g.shape[1], (
+            "`g` should be a square adjacency matrix"
+        )
+        self.g = self.g.astype(np.float64, copy=False)
+        self._vals = _resolve_vals(vals)
```

Comment:
I think maybe something like `data` is better than `vals` which us "`_SparseMetric` operates on its `graph` and `data`" instead of "`_SparseMetric` operates on its `g` and `_vals`"

Separately, this `_resolve_val` should be called by the caller IMO and we can use the `__post_init__` to validate and raise `TypeError`.  Code duplication in exchange for not having `vals` (which is never used) and `_vals` is worth it IMO

Reply:
`vals` is used in the 1D case, but good point!

Reply:
If it is then I don't see it, maybe I'm missing something.

https://github.com/scverse/scanpy/pull/3553/files/c4e037f9a36f9618a1fe3fb3c2d24f17f261b226..9f0f18589fe81c389c8b3072b2331bd0f79f446f#diff-fb08f0d4e96aca4f02fb2014b003a0e560fcddff14bbc532428f407ef5d1aaabR125

https://github.com/scverse/scanpy/pull/3553/files/c4e037f9a36f9618a1fe3fb3c2d24f17f261b226..9f0f18589fe81c389c8b3072b2331bd0f79f446f#diff-e6b16594f2d52a49a1e95379f8387e9c8951396a2f34eb51f12d81326eb6a67bR123

Reply:
https://github.com/scverse/scanpy/blob/f5ef8f76f4e31d2f5eceb3f007f594811d5099ab/src/scanpy/metrics/_gearys_c.py#L125